### ISSUE TITLE
Restore doesn't work for upgrading iPhone 8 Plus 14.8 -> 15.1 beta 3 in DFU mode

### ISSUE THREAD

**Issue by ChrisCheney** [doronz88/pymobiledevice3#156]
**Test environment**

Fedora 34
iPhone 8 Plus - iOS 14.8 -> 15.1 beta 3

**Describe the bug**
Trying to upgrade from iOS 14.8 to 15.1 beta 3 running restore while in DFU pymobildevice3 gets to the point that it reboots the phone the first time and then the phone goes back into iOS GUI instead of the recovery mode.

**To Reproduce**
1. Put phone in DFU mode
2. Run pymobiledevice3 restore update --erase iPhone_5.5_P3_15.1_19B5060d_Restore.ipsw 
3. Watch it reboot back into iOS GUI instead of proceeding to the next step after the first reboot
4. See logs

**Expected behavior**
It should update, but does not.

**Logs**
```
# pymobiledevice3 restore update --erase iPhone_5.5_P3_15.1_19B5060d_Restore.ipsw 
2021-10-07 22:43:35 HOSTNAME root[8266] DEBUG searching among connected devices via lockdownd
2021-10-07 22:43:35 HOSTNAME root[8266] DEBUG waiting for device to be available in Recovery mode
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG set_configuration: 1
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG set_interface_altsetting: 0 0
2021-10-07 22:43:36 HOSTNAME root[8266] INFO connected device: <ecid: XXXXXXXXXXXXXXXX hardware_model: d21ap image4-support: 4>
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG scanning BuildManifest.plist for the correct BuildIdentity
2021-10-07 22:43:36 HOSTNAME root[8266] INFO fetching TSS record
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry AOP
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry AVE
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry Ap,SystemVolumeCanonicalMetadata
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry AppleLogo
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry AudioCodecFirmware
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryCharging0
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryCharging1
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryFull
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryLow0
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryLow1
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry BatteryPlugin
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry DeviceTree
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry ISP
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry KernelCache
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry LLB
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry Liquid
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry Multitouch
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry OS
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RecoveryMode
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreDeviceTree
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreKernelCache
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreLogo
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreRamDisk
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreSEP
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry RestoreTrustCache
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry SEP
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry StaticTrustCache
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry SystemVolume
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry ftap
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry ftsp
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry iBEC
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry iBSS
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry iBoot
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry rfta
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Applying restore request rules for entry rfts
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding ESEC=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG Adding EPRO=True to TSS entry
2021-10-07 22:43:36 HOSTNAME root[8266] INFO Sending TSS request...
2021-10-07 22:43:36 HOSTNAME urllib3.connectionpool[8266] DEBUG Starting new HTTP connection (1): gs.apple.com:80
2021-10-07 22:43:36 HOSTNAME urllib3.connectionpool[8266] DEBUG http://gs.apple.com:80 "POST /TSS/controller?action=2 HTTP/1.1" 200 None
2021-10-07 22:43:36 HOSTNAME root[8266] INFO response successfully received
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG NOTE: No path for component iBSS in TSS, will fetch from build_identity
2021-10-07 22:43:36 HOSTNAME root[8266] INFO Personalizing IMG4 component iBSS...
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG tag: Tag(nr=22, typ=0, cls=0) ibss
2021-10-07 22:43:36 HOSTNAME root[8266] DEBUG irecv state: 2
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 528/528 [00:00<00:00, 799.36it/s]
2021-10-07 22:43:37 HOSTNAME root[8266] DEBUG waiting for status == 5
2021-10-07 22:43:37 HOSTNAME root[8266] DEBUG resetting usb device
```

**Comment by ChrisCheney** [comment]
Switching over to recovery mode instead of DFU mode seemed to work but it got one exception during the process.
```
2021-10-07 22:58:24 HOSTNAME root[12156] INFO about to send filesystem...
2021-10-07 22:58:24 HOSTNAME root[12156] DEBUG connecting to ASR
2021-10-07 22:58:24 HOSTNAME root[12156] DEBUG ASR connected
2021-10-07 22:58:24 HOSTNAME root[12156] DEBUG got command: {'Checksum Chunks': True, 'Command': 'Initiate'}
2021-10-07 22:58:24 HOSTNAME root[12156] INFO connected to ASR
2021-10-07 22:58:24 HOSTNAME root[12156] INFO validating the filesystem
2021-10-07 22:58:50 HOSTNAME root[12156] DEBUG FDR <pymobiledevice3.restore.fdr.FDRClient object at 0x7fb777b6d790> got payload of 31 bytes, now try to proxy it
2021-10-07 22:58:50 HOSTNAME root[12156] DEBUG Sending 31 bytes of data
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/root/pymobiledevice3/pymobiledevice3/restore/fdr.py", line 196, in fdr_listener_thread
    client.poll_and_handle_message()
  File "/root/pymobiledevice3/pymobiledevice3/restore/fdr.py", line 184, in poll_and_handle_message
    handlers[cmd]()
  File "/root/pymobiledevice3/pymobiledevice3/restore/fdr.py", line 117, in handle_proxy_cmd
    return self.poll_and_handle_message()
  File "/root/pymobiledevice3/pymobiledevice3/restore/fdr.py", line 184, in poll_and_handle_message
    handlers[cmd]()
  File "/root/pymobiledevice3/pymobiledevice3/restore/fdr.py", line 162, in handle_proxy_cmd
    self.service.sendall(buf)
  File "/root/pymobiledevice3/pymobiledevice3/service_connection.py", line 61, in sendall
    self.socket.sendall(data)
BrokenPipeError: [Errno 32] Broken pipe

2021-10-07 22:59:57 HOSTNAME root[12156] INFO filesystem validated
2021-10-07 22:59:57 HOSTNAME root[12156] INFO sending filesystem now...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 42431/42431 [02:35<00:00, 273.23it/s]
verify-restore: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:09<00:00, 10.41it/s]
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': True, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1627, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': False, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1664, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': True, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1664, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': False, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1653, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': True, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1653, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': False, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1676, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': True, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1676, 'MsgType': 'CheckpointMsg'}
2021-10-07 23:02:48 HOSTNAME root[12156] INFO checkpoint: {'CHECKPOINT_COMPLETE': False, 'CHECKPOINT_RESULT': 0, 'CHECKPOINT_ID': 1574, 'MsgType': 'CheckpointMsg'}
```

**Comment by doronz88** [comment]
Thanks for the report! I didn't test updating from DFU. I'll test it and work out a fix on the next few days.

The FDR exception isn't really terrible. It's mainly for apple to retireve the crash reports during the update process. I'll try to figure out what went wrong  there.

**Comment by doronz88** [comment]
So far I haven't been able to reproduce the DFU update issue. Tested iPhone 8 Plus 15.0 -> 15.0.1 (Apple stopped signing 14.8). Did this occur every time or was it maybe some rare race during the update? @ChrisCheney

**Comment by doronz88** [comment]
I tried to resolve the FDR issue on #157. I'm not sure since I wasn't able to reproduce this bug also, but I believe it should fix it.

**Comment by ChrisCheney** [comment]
The DFU issue appears to happen every time and happened on two separate iPhone 8 Plus units.

The update you made for #157 appears to have not fixed the error message that is shown and I got three of them on the other unit when I updated to 15.1 beta 3 with current git. But it still seemed to work for the upgrade.

**Comment by doronz88** [comment]
Ok, it's actually very interesting. I was able to reproduce this every time updating to iOS 15.1 beta X, but updating to iOS 15.0.1 seems to not have this bug. 

**TL;DR**
This bug occurs also when updating from iTunes, meaning it's on Apple's end. Nothing we can/should do :).

**Further explenation on what went wrong**

- Usually, when the device enters recovery mode, a variable named `autoboot` is set to False.
- When trying to reboot with `autoboot = False` the device will always enter recovery mode.
- Setting this value to True and sending the reboot command makes it reboot to normal mode.
- From some reason, I guess that sending the iBSS of iOS 15 Beta X (new recovery image) doesn't set this variable to False, allowing the device to boot normally.

The good news is that entering recovery won't cause this bug, since the enter to recovery sets this variable correctly and doesn't require the send of the iBSS payload which causes it to boot to normal once more.


About the FDR issue, I really wasn't able to reproduce it, only when simulating network issues. It's not severe at all as I have previously mentioned since for my understanding it's only meant for crash reporting. Can you maybe attach the full FDR log? mainly the part of:

https://github.com/doronz88/pymobiledevice3/blob/master/pymobiledevice3/restore/fdr.py#L132
